lib/logstash/string_interpolation.rb

```diff
@@ -0,0 +1,155 @@
+      if template.is_a?(Float) and
+          (template < MIN_FLOAT_BEFORE_SCI_NOT or template >= MAX_FLOAT_BEFORE_SCI_NOT) then
+        return template = ("%.15f" % template).sub(/0*$/,"")
+      end
+      
+      template = template.to_s
+
+      if not_cachable?(template)
+        return template
+      else
+        compiled = CACHE.get_or_default(template, nil) || CACHE.put(template, compile_template(template))
+        compiled.evaluate(event)
+      end
+    end
+
+    def self.evaluate(event, format)
+      INSTANCE.evaluate(event, format)
+    end
+
+    private
+    def not_cachable?(template)
+      template.index("%").nil?
+    end
+
+    def compile_template(template)
+      nodes = Template.new
+
+      position = 0
+      matches = template.to_enum(:scan, TEMPLATE_TAG_REGEXP).map { |m| $~ }
+
+      if matches 
+        matches.each do |match|
+          tag = match[0][2..-2]
+          start = match.offset(0).first
+          nodes << StaticNode.new(template[position..(start-1)]) if start > 0
+          nodes << identify(tag)
+          position = match.offset(0).last
+        end
+      end
+
+      if position < template.size - 1
+        nodes << StaticNode.new(template[position..-1])
+      end
+
+      optimize(nodes)
+    end
+
+    def optimize(nodes)
+      if nodes.size == 1
+        nodes.first 
+      else 
+        nodes
+      end
+    end
+
+    def identify(tag)
+      if tag == "+%s"
+        EpocNode.new
+      elsif tag[0, 1] == "+"
+        DateNode.new(tag[1..-1])
+      else
+        KeyNode.new(tag)
+      end
+    end
+  end
+
+  class Template
+    extend Forwardable
+    def_delegators :@nodes, :<<, :push, :size, :first
+
+    def initialize
+      @nodes = []
+    end
+
+    def evaluate(event)
+      @nodes.collect { |node| node.evaluate(event) }.join
+    end
+  end
+
+  class EpocNode
+    def evaluate(event)
+      t = event.timestamp
+      if t.nil?
+        raise LogStash::Error, "Unable to format in string \"#{@format}\", #{LogStash::Event::TIMESTAMP} field not found"
+      end
+      t.to_i.to_s
+    end
+  end
+
+  class StaticNode
+    def initialize(content)
+      @content = content
+    end
+
+    def evaluate(event)
+      @content
+    end
+  end
+
+  class KeyNode
+    def initialize(key)
+      @key = key
+    end
+
+    def evaluate(event)
+      value = event[@key]
+
+      case value
+      when nil
+        "%{#{@key}}"
+      when Array
+        value.join(",")
+      when Hash
+        Logstash::Json.dump(value)
+      else
+        value
+      end
+    end
+  end
+
+  class DateNode
+    def initialize(format)
+      @format = format
+      @formatter = org.joda.time.format.DateTimeFormat.forPattern(@format)
+          .withZone(org.joda.time.DateTimeZone::UTC)
+    end
+
+    def evaluate(event)
+      t = event.timestamp
+
+      if t.nil?
+        raise LogStash::Error, "Unable to format in string \"#{@format}\", #{LogStash::Event::TIMESTAMP} field not found"
+      end
```

Comment:
if the line is short < 100, I use `.... if @test` When it's too long I prefer to use it first, so when I read the code my brain just skip it without reading the line.

I have a hard limit on `width` on my vim setup even in 2015.. :monkey: